### DISCUSSION TITLE
[@types/aws-lambda] can we get rid of all the `null`s?

### DISCUSSION THREAD

**Discussion by G-Rath** [DefinitelyTyped/DefinitelyTyped#69886] [Issues with a @types package]
I'm hoping to get time to dig into this sometime this year, but for now I thought I'd kick off a discussion to capture my thinking and so others can weight in or even pick this up if they've got the bandwidth.

Right now there's a lot of properties marked with `| null` for things that I expect to always be present in the real-world but that technically you can omit (or explicitly set to `null`?) when using test events from the console.

For example:

```
export interface APIGatewayProxyEventBase<TAuthorizerContext> {
    body: string | null;
    headers: APIGatewayProxyEventHeaders;
    multiValueHeaders: APIGatewayProxyEventMultiValueHeaders;
    httpMethod: string;
    isBase64Encoded: boolean;
    path: string;
    pathParameters: APIGatewayProxyEventPathParameters | null;
    queryStringParameters: APIGatewayProxyEventQueryStringParameters | null;
    multiValueQueryStringParameters: APIGatewayProxyEventMultiValueQueryStringParameters | null;
    stageVariables: APIGatewayProxyEventStageVariables | null;
    requestContext: APIGatewayEventRequestContextWithAuthorizer<TAuthorizerContext>;
    resource: string;
}
```

As far as I know, API Gateway will always provide query and path parameter values even if they're just empty.

What I'm wondering is if we could introduce a "settings" type like what `Luxon` has that would let people choose to opt-in/out to the `null` types.

For Luxon this works for the handling of invalid dates - by default, Luxon returns an object with invalid properties (`invalidReason`, `invalidExplanation` etc) + null properties if a date is invalid, but you can set it to throw if a date is invalid in which case you know that common properties will never be `null`.

This is enabled at runtime with the following:

```
import { Settings } from 'luxon';

Settings.throwOnInvalid = true;
```

For TypeScript, they have this cool [`TSSettings`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/luxon/src/settings.d.ts) interface that you can use in a similar way:

```
declare module 'luxon' {
  interface TSSettings {
    throwOnInvalid: true;
  }
}
```

This changes the value of [this `CanBeInvalid` utility type](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/luxon/src/_util.d.ts#L3), which can be used in turn to switch [between properties having different values](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/luxon/src/datetime.d.ts#L896)

So I'm wondering if we could do something similar:

```
declare module 'aws-lambda' {
  interface TSSettings {
    // todo: better name pending
    iWillNeverUseTestEvents: true;
  }
}

export type SupportsTestEvents = TSSettings extends { iWillNeverUseTestEvents: true } ? false : true;

export type CanBeNullWhenTesting<ValidType> = SupportsTestEvents extends true
    ? ValidType | null
    : ValidType

export interface APIGatewayProxyEventBase<TAuthorizerContext> {
    body: string | null;
    headers: APIGatewayProxyEventHeaders;
    multiValueHeaders: APIGatewayProxyEventMultiValueHeaders;
    httpMethod: string;
    isBase64Encoded: boolean;
    path: string;
    pathParameters: CanBeNullWhenTesting<APIGatewayProxyEventPathParameters>;
    queryStringParameters: CanBeNullWhenTesting<APIGatewayProxyEventQueryStringParameters>;
    multiValueQueryStringParameters: CanBeNullWhenTesting<APIGatewayProxyEventMultiValueQueryStringParameters>;
    stageVariables: CanBeNullWhenTesting<APIGatewayProxyEventStageVariables>;
    requestContext: APIGatewayEventRequestContextWithAuthorizer<TAuthorizerContext>;
    resource: string;
}
```

[a comment by an automated account is not shown]

**Comment by simonbuchan** [comment]
I like the idea, but seems like it belongs in a major version update (even if it's strictly a compatible change)

Making it ok to take major version bumps (eg making compatability changes in one trigger) is a bit of a problem, really, I picked the runtime interface changing when I did it last, but it seems we won't be getting any of those again soon.

Maybe if I ever get off my butt and break this uber-package into specific ones that are fine to bump (think `@aws-lambda/runtime`, `@aws-lambda/trigger-api-gateway`, or something)

**Reply by darbio** [reply]
> Maybe if I ever get off my butt and break this uber-package into specific ones that are fine to bump (think @aws-lambda/runtime, @aws-lambda/trigger-api-gateway, or something)

I think this is the way we should be headed. I'm also wondering if we should move away from DT so that we can add in some of the more function-y changes that make sense and get requested time and time again.

**Reply by simonbuchan** [reply]
You mean like actual runtime helper functions? 

I would say I would feel a little bit skeezy taking a package a bunch of people have contributed to in a place where anyone can contribute even if we all disappear and moving it to a "private" repo - and it seems like it would be a bit annoying to set up all the testing and publishing workflows to the same level.

Maybe a middle-ground would be to reserve an npm org for the packages where the typing packages still come from here but the functional helpers are from a separate repo?

**Reply by darbio** [reply]
> You mean like actual runtime helper functions?

Yeah. Something which we can use to address the things that "don't make sense", things that are requested time and time again, and which stops people from using the wrong handler types. 

> I would say I would feel a little bit skeezy taking a package a bunch of people have contributed to in a place where anyone can contribute even if we all disappear and moving it to a "private" repo. 

I think this is something that we can discuss and see if there is an appetite for it. The licensing is pretty permissive and would allow for a fork. 

GitHub organisations can handle this pretty well and remain open and transparent. Anyone can still contribute and maintain the library. 

> and it seems like it would be a bit annoying to set up all the testing and publishing workflows to the same level.

This is pretty trivial using GitHub. I think we could even use the DT CI/CD workflow. 

> Maybe a middle-ground would be to reserve an npm org for the packages where the typing packages still come from here but the functional helpers are from a separate repo?

Yeah that's not a bad idea, however it does tie the functional helpers to DT types. This isn't a particular blocking thing, but it could get messy. 

I've also been a bit annoyed recently by the amount of time that it's taken for a DT maintainer to get on the approved requests - I understand that it's due to the popularity of these definitions, but it's taking longer and longer to get them into the master. 

I've maintained from the beginning that AWS themselves should be backing this with assistance to get it right.

SDK V3 went a long way but not far enough to make these typings obsolete. Perhaps the solution is for AWS to create an @aws-sdk/lambda-server package which included appropriate types. 

Just food for though - I'm not 100% convinced either way.
